File: syllabify.py

EN_VOWELS = {"AA","AE","AH", "AO","AW","AX","AXR","AY","EH","ER","EY","IH","IX","IY","OW","OY","UH","UW","UX"}
import re

def get_words_from_file(fname: str):
    '''
    Reads a wordlist file that contains one word per line and returns a list of the words
    '''
    with open(fname, 'r') as f:
        word_list = [word.lower().strip() for word in f]
    return word_list


def get_transcription_dict_from_file(fname: str):
    '''
    Reads CMU dict ARPABET transcription file into a dictionary and returns it
    '''
    token_dict = {}
    with open(fname, 'r') as f:
        for line in f:
            line = line.strip()
            token_list = line.split(" ")
            token_dict[token_list[0]] = tuple(token_list[1:])
    return token_dict

def lookup_transcriptions(wordlist: list, transcriptiondict: dict):
    '''
    Attempts to match transcriptions to items in wordlist, returns a dictionayr of transcribed words or a list of out of vocabulary words
    '''
    transcribed = {word:transcriptiondict[word] for word in wordlist if word in transcriptiondict.keys()}
    OOVwords = [word for word in wordlist if word not in transcriptiondict.keys()]
    return transcribed, OOVwords
        

def estimate_onsets(transcriptions: dict[str:tup[str]], vowels: set[str]):
    '''
    Find all the attested word-initial onsets and returns them as a list of all licit onsets according to the dataset
    '''
    licit_onsets = set()
    for pronunciation in transcriptions.values():
        for i, phone in enumerate(pronunciation):
            vowel = re.sub(r"\d", "", phone)
            if vowel in vowels:
                tup = tuple(list(pronunciation)[:i])
                if(len(tup) != 0):
                    licit_onsets.add(tup)
                break
    return licit_onsets


def mark_vowels(transcription: tup[str], vowels: set[str]):
    '''
    Find the locations of each vowel in the transcription and returns it as a sequence of integers corresponding to indices of the vowels.
    '''
    indexes = [ i for i, phone in enumerate(transcription) if (re.sub(r"\d", "", phone)) in vowels]
    return indexes


def mark_syllstarts(transcription: tup[str], nuc_indices: list[int], licit_onsets: list[str[):
    '''
    Find the locations of the first phone of each syllable and returns it as a sequence of integers corresponding to indices of syllable starts.
    '''
    syllable_locations = []
    if len(nuc_indices) == 1:
        return []
    n = -2
    while(-len(nuc_indices) <= n):
        prev = nuc_indices[n]
        next = nuc_indices[n+1]
        candidate = list(transcription)[prev + 1:next]
        while (tuple(candidate) not in licit_onsets) and len(candidate) > 0:
            if len(candidate) == 1:
                candidate = []
            else:
                candidate = candidate[1:]
            prev+=1
        syllable_locations.append(prev)
        n -=1
    syllable_locations.sort()
    return syllable_locations


def split_transcription(transcription: list[str], syllstart_indices: list[int]):
    '''
    Splits a transcription into a sequence of syllables and returns the list of strings representing each syllable
    '''
    transcription = list(transcription)
    syllstart_indices.append(10**10) # This will prevent the list comprehension from breaking
    syllstart_indices.append(-1)
    syllstart_indices.sort()
    syllable_list = [tuple(transcription[syllstart_indices[i] + 1: min(syllstart_indices[i + 1] + 1, len(transcription))]) for i in range(len(syllstart_indices) - 1)]
    # the above list comprehension slices transcriptions between the indices of syllstart. I added -1 and +infinity so that i can control the situation where the slice is in the start,
    # or at the end
    return syllable_list


def get_syllabifications(transcriptions: list[str], licit_onsets:list[str], vowels:list[str]):
    '''
    Syllabify a series of transcriptions and return a dictionary of syllabified transcriptions.
    '''
    syllabifications = {}
    for word, trans in transcriptions.items():
        nucleus_indices = mark_vowels(trans, vowels)
        syllstart_indices = mark_syllstarts(trans, nucleus_indices, licit_onsets)
        syllabifications[word] = split_transcription(trans, syllstart_indices)
    return syllabifications


def write_syllables(fname: str, syllabified_dict: dict):
    '''
    Writes syllabifications to a file in a specific format. Each line in the output file contains a tab-separated word-syllabification pair and ends in a newline.
    The syllabification is in ARPABET notation, but with a . inserted between syllables.
    '''
    with open(fname, 'w') as f:
        for word, trans in syllabified_dict.items():
            for i, tup in enumerate(trans):
                joined = " ".join(tup)
                trans[i] = joined
            syllables = " . ".join(trans)
            line = f"{word}\t{syllables}\n"
            f.write(line)
            
def syllabify_english(wordlistfname: str, outfname: str):
    '''
    This function looks up transcriptions for a word list and syllabifies them
    '''
    vowels = EN_VOWELS # collection of vowels for English ARPABET

    wordlist = get_words_from_file(wordlistfname)
    cmudict = get_transcription_dict_from_file("cmudict.dict")
    
    transcribed_wordlist, OOV_wordlist = lookup_transcriptions(wordlist, cmudict)
    licit_onsets = estimate_onsets(transcribed_wordlist, vowels)
    syllabified_dict = get_syllabifications(transcribed_wordlist, licit_onsets, vowels)
    write_syllables(outfname, syllabified_dict)



def main():

    infname = "wordlist.txt"
    outfname = "syllabified.txt"
    syllabify_english(infname, outfname)


if __name__ == "__main__":
    main()

File: syllabify.mjs
/**The following javascript program is a maxmum onset principle based automatic syllabifier of english text. The algorithm takes in phone data from the provided corpus and uses this as a basis for estimating the syllable boundaries of each word. then it writes this into an output text file. */
import dictionary from './en_us_arpa.json' with {type: 'json'};
import onsets from './onset_database.json' with {type: 'json'};

/**
*runs the main program of the syllabifier
*@param string[] this is a filtered line by line representation of the user's input.
*@returns {string} the correct syllabification of each of the user's words (if they are in the dictionary) each printed on one line exactly separated by a tab from their syllabifications.
*/
function syllabify(lines){
	console.log(`received lines in syllabifier ${lines} type of lines ${typeof lines}`);
	const enVowels = new Set(["AA","AE","AH", "AO","AW","AX","AXR","AY","EH","ER","EY","IH","IX","IY","OW","OY","UH","UW","UX"]);
	console.log(`obtaining a word list`);
	const wordList = getWordList(lines);
	console.log(`obtained a word list ${[wordList, typeof wordList]}`);
	const [transcribedWordlist, oovList ] = lookupTranscriptions(wordList, dictionary);
	console.log(`syllabifiying the transcriptions ${[transcribedWordlist, typeof transcribedWordlist]}`);
	const syllabified = getSyllabifications(transcribedWordlist, onsets, enVowels);
	return syllabified;
	}

/**
* Takes the list of transcriptions and applies the maximum onset principle to them. Then returns a string representing the syllabified transcription of the input.
*@param transcribedWordlist string[] a list of strings that are the utterances of the user in their phonemic form
*@param onsets string[] a list of strings that are considered valid beggingings of a vowel center.
*@param enVowels {Set} a set of strings representing the vowel sounds of the langugage.
*/
function getSyllabifications(transcribedWordlist, onsets, enVowels){
	const syllabified = [];
	for (const word of transcribedWordlist){
		if (word === '<unknown>'){
			syllabified.push(word);
			continue;
			}
		console.log(`marking vowels ${[word, typeof word]}`);
		const nucleusIndices = markVowels(word, enVowels);
		console.log(`marking syllable starts`);
		const syllStartIndices = markSyllStarts(word, nucleusIndices, onsets);
		const syllabification = splitTranscription(word, syllStartIndices);
		syllabified.push(syllabification.join(" . "));
		}
	console.log(`output to be sent ${syllabified}`);
	return syllabified.join(" | ");
	}

/**
*  Find the locations of each vowel in the transcription and returns it as a sequence of integers corresponding to indices of the vowels.
*@param word {string} a phonemic sequence
*@param enVowels {Set} a set of english vowels
*@returns int[] a list of indices marking the location of the vowel
*/
function markVowels(word, enVowels){
	const array = []
	const transcription = word.split(" ");
	for (const i in transcription){
		var phone = transcription[i].replaceAll(/\d/g, "");
		console.log(`phone found ${[phone, typeof phone]}`);
		if (enVowels.has(phone)){
			console.log(`vowel found`);
			array.push(parseInt(i));
			}
		}
	console.log(`marked vowel index ${array}`);
	return array;
	}

/**
* finds the locations of the first phone of each syllable in the transcription and returns it as a sequence of integers corresponding to indices of syllable starts.
*@param word {string} the transcription of the utternace
*@param nucleusIndices int[] identifies where the vowel phone is in the word.
*@param onsets string[] a list of valid phone sequences
*@returns int[] the starting location of each syllable in the word
*/
function markSyllStarts(word, nucleusIndices, onsets){
	const licitOnsets = new Set(onsets);
	const transcription = word.split(" ");
	console.log(`working with transcription ${transcription}`);
	const syllableLocations = [];
	console.log(`nucleus indices ${nucleusIndices}`);
	if (nucleusIndices.length === 1){ return [];}
	var n = -2;
	//the loop processes the string in reverse order since onsets are by default a prefix of the syllable in question. 
	while(-(nucleusIndices.length) <= n){
		var prev = nucleusIndices.at(n);
		var next = nucleusIndices.at(n+1);
		// get the largest possible onset then sequentially shrink it based on whether or not it is valid
		console.log(`slicing between ${prev + 1} and ${next}`);
		var candidate = transcription.slice(prev + 1, next);
		console.log(`onset candidate ${candidate}`);
		while (!licitOnsets.has(candidate.join(" ")) && candidate.length > 0){
			if (candidate.length === 1){
				candidate = [];
				}
			else{
				candidate = candidate.slice(1);
				console.log(`new candidate ${candidate}`);
				prev+=1;
				
				}
			}
		console.log(`syllable start index found ${prev}`);
		syllableLocations.push(prev);
		n -=1;
		}
	syllableLocations.sort();
	console.log(`marked syllable starts ${syllableLocations}`);
	return syllableLocations;
	}


/**
*given the starting position of each syllable in this word, it chunks the phones by syllable then returns them as a list.
*@param word {string} a phonemic sequence
*@param syllStartIndices int[] the starts of each syllable in word.
*@returns string[] a list of strings representing the syllables of this word
*/
function splitTranscription(word, syllStartIndices){
	const transcription = word.split(" ");
	syllStartIndices.push(9999);
	syllStartIndices.push(-1);
	syllStartIndices.sort();
	const syllableList = [];
	console.log(`splitting transcription by syllables ${transcription} with locations ${syllStartIndices}`);
	for (var i = 0; i < syllStartIndices.length - 1; i +=1){
		console.log(`slicing between ${syllStartIndices[i] + 1} and ${Math.min(syllStartIndices[i + 1] + 1, transcription.length)}`);
		var syllable = transcription.slice(syllStartIndices[i] + 1, Math.min(syllStartIndices[i + 1] + 1, transcription.length)).join(" ");
		console.log(`syllable found ${syllable}`);
		syllableList.push(syllable);
		}
	return syllableList;
	}
/**
*   Attempts to match the words with their pronunciations returns a dictionary of transcribed words and a list of out of vocabulary words
*@param string[] a list of words representing the user's input
*@param dictionary {string: string} is a dictionary that maps a script representation of a word to the phoneme representation.
* @returns list[] the object contains two lists, one is a list of transcriptions for each key, and the other is a list of invalid keys. 
*/
function lookupTranscriptions(wordList, dictionary){
	const transcribedWordlist = [];
	const oovList = [];
	for (const word of wordList){
		if (typeof word === "string"){	
			var value = dictionary[word];
			if (typeof value === "string"){
				transcribedWordlist.push(value);
				}
			else{
				transcribedWordlist.push('<unknown>');
				oovList.push(word);
				}
			
			}
		}
	console.log(`generated the transcriptions ${transcribedWordlist}`);
	return [transcribedWordlist, oovList];
	}


/**
*reads each line in the input and breaks it down by whitespace, extracting the token contained. it returns the result as a list of words
*@param string[] lines a list of lines written by the user and filtered
*@returns string[] a list of words that the user wrote.
*/
function getWordList(lines){
	var wordList = [];
	for (const line of lines){
		console.log(`processing line ${[line, typeof line]}`);
		var tokens = strip(line).split(" ");
		var array = [];
		for (const token of tokens){
			console.log(`token extracted ${[token, typeof token]}`);
			array.push(strip(token));
			}
		wordList = wordList.concat(array);
		}
	console.log(`created a wordlist ${wordList}`);
	return wordList;
	}
	
/**
*strips whitespace from the ends of the string
*@param {string} the string to be stripped
*@returns {string} a stripped string
*/
function strip(string){
	console.log(`stripping line ${[string, typeof string]}`);
	var tokens = string;
	if (tokens[-1] === " "){
		tokens = tokens.slice(0, tokens.length - 1);
		}
	if (tokens[0] === " "){
		tokens = tokens.slice(1);
		}
	console.log(`stripped line ${[tokens, typeof tokens]}`);
	return tokens;
	}
	
export {syllabify, strip};
	
	
	
	
	
	
	
	
	
	
	